### PR TITLE
Rewrite specs using rspec

### PR DESCRIPTION
**Pull request by kbogtob**
### What does this PR?

It rewrites the tests using RSpec and mocking in the idea of refactoring the project to have more SOLID code. Some tests are still breaking encapsulation and fixing them implies to modify the code to make it more testable and uncoupled. 

### Motivation

Clarify the project, use more recent testing techniques and allowing people to contribute easily + making grounds to refactor the project.

### Notes

Found a bug, notified it in comment.

### CODE REVIEW

File: lib/datadog/statsd/uds_connection.rb [hunk 1]
```diff
@@ -28,6 +28,9 @@ def send_message(message)
         socket.sendmsg_nonblock(message)
       rescue Errno::ECONNREFUSED, Errno::ECONNRESET, Errno::ENOENT => e
         @socket = nil
+        # TODO: FIXME: This error should be considered as a retryable error in the
```
Comment (kbogtob) [inline lib/datadog/statsd/uds_connection.rb]: I found this bug, it was not caught by the previous tests: If we catch a connection refused or reset, we don't try to reconnect to the UDS socket. For the missing socket, it seems relevant, it means that the socket was never found, but for other cases, I think we should retry.

File: spec/statsd/version_spec.rb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+require 'spec_helper'
+
+describe Datadog::Statsd do
+  describe 'VERSION' do
+    it 'has a version' do
+      expect(Datadog::Statsd::VERSION).to eq '4.6.0'
```
[the remark this answers is not in the document]
Comment (kbogtob) [reply spec/statsd/version_spec.rb]: I like this part to be explicit everytime we release but I can change it to a regexp

File: spec/matchers/telemetry_matcher.rb [hunk 3]
```diff
@@ -0,0 +1,26 @@
+require 'rspec/expectations'
+
+RSpec::Matchers.define :eq_with_telemetry do |expected, telemetry_options|
+  telemetry_options ||= {}
+
+  def text_with_telemetry(text, metrics: 1, events: 0, service_checks: 0, bytes_sent: 0, bytes_dropped:0, packets_sent: 0, packets_dropped: 0, transport: 'udp')
```
[the remark this answers is not in the document]
Comment (kbogtob) [reply spec/matchers/telemetry_matcher.rb]: It is!

File: spec/integrations/allocation_spec.rb [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        namespace: namespace,
+        sample_rate: sample_rate,
+        tags: tags,
+        logger: logger,
+        disable_telemetry: true,
+      )
+    end
+
+    let(:expected_allocations) do
+      if RUBY_VERSION >= '2.3.0' && RUBY_VERSION < '2.4.0'
```
[the remark this answers is not in the document]
Comment (kbogtob) [reply spec/integrations/allocation_spec.rb]: It is what I wanted :). Older rubies make 8 allocations in this case.

File: spec/integrations/allocation_spec.rb [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+  before do
+    allow(Socket).to receive(:new).and_return(socket)
+    allow(UDPSocket).to receive(:new).and_return(socket)
+    # initializing statsd so it does not count in allocations
+    subject.increment('foobar')
+  end
+
+  let(:expected_allocations) do
+    if RUBY_VERSION >= '2.3.0' && RUBY_VERSION < '2.4.0'
```
[the remark this answers is not in the document]
Comment (kbogtob) [reply spec/integrations/allocation_spec.rb]: It is what I wanted :). Older rubies make 8 allocations in this case.

File: lib/datadog/statsd/version.rb [hunk 6]
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+require_relative 'connection'
+
+module Datadog
+  class Statsd
+    VERSION = '4.6.0'
```
Comment (remeh) [inline lib/datadog/statsd/version.rb]: Nit: it's most likely `4.7.0` now.

File: spec/statsd/version_spec.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (remeh) [reply spec/statsd/version_spec.rb]: Let's properly update the release documentation then, let me know if you need the link to the document I'm mentioning

File: spec/integrations/connection_edge_case_spec.rb [hunk 7]
```diff
@@ -0,0 +1,467 @@
...
+              subject.write('foobar')
+            end.not_to raise_error
+          end
+
+          it 'logs the error message' do
+            subject.write('foobar')
+            expect(log.string).to match 'Statsd: SocketError yolo'
+          end
+        end
+      end
```
Comment (remeh) [inline spec/integrations/connection_edge_case_spec.rb]: Aren't they testing the same thing? I'm may miss a subtlety here.

File: spec/integrations/live_testing_spec.rb [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+  context 'with a real UDP socket' do
+    it 'should actually send stuff over the socket' do
+      socket = UDPSocket.new
+      host, port = 'localhost', 12345
+      socket.bind(host, port)
+
+      statsd = Datadog::Statsd.new(host, port)
+      statsd.increment('foobar')
+      message = socket.recvfrom(64).first
+      expect(message).to eq 'foobar:1|c'
```
Comment (remeh) [inline spec/integrations/live_testing_spec.rb]: Should we test more than just a counter? Probably not has you improved the test coverage on everything but if that's like 2, 4 more lines, could be helpful someday.

File: spec/statsd_spec.rb [hunk 9]
```diff
@@ -1,919 +1,856 @@
...
+      logger: logger,
+      telemetry_flush_interval: -1,
+    )
   end
 
-  let(:namespace) { nil }
+  let(:namespace) { 'sample_ns' }
   let(:sample_rate) { nil }
-  let(:socket) { FakeUDPSocket.new }
+  let(:tags) { %w[abc def] }
```
Comment (remeh) [inline spec/statsd_spec.rb]: Because it doesn't harm, let's use tags just slightly more realistic: for instance `abc:def host:myhost12-11`, WDYT?

Comment (remeh) [reply spec/statsd_spec.rb]: Actually you did it right after, nvm.

File: lib/datadog/statsd/version.rb (hunk 6 above)
Comment (kbogtob) [reply lib/datadog/statsd/version.rb]: I'll fix it :)

File: spec/integrations/connection_edge_case_spec.rb (hunk 7 above)
Comment (kbogtob) [reply spec/integrations/connection_edge_case_spec.rb]: It is, just to be sure that we don't handle socketerrors differently.

File: spec/integrations/live_testing_spec.rb (hunk 8 above)
Comment (kbogtob) [reply spec/integrations/live_testing_spec.rb]: We could make a big test for that. I'll add it.

**Pull request merged**